### PR TITLE
Fix inconsistent URL filter matching across campaign decisions and points

### LINKED ISSUE #12087: UX bug: Mautic uses different ways to filter URLs across the application: campaign decisions and points
### Mautic Version

4.4.x series

### PHP version

7.4

### What browsers are you seeing the problem on?

Not relevant

### What happened?

Depends on the place in the app Mautic uses a different logic to filter URLs. 

Some use a [shell wildcard pattern](https://www.php.net/manual/en/function.fnmatch.php) like `*test*`
- Points
- Visited a page

Some use [regular expressions](https://www.php.net/manual/en/function.preg-match.php) like `test`
- Clicks a URL in an email

So if I want to search for a string 'test' inside a URL I have to use `*test*` or `test` depending what feature of Mautic I'm using. 

This should be consolidated into one. 

IMO it should be all regular expressions. This is what people are used from e.g. Google Analytics, and it does not need any wildcards for a standard string in string search case. See the example above.

We can use the `UrlMatcher::hasMatch` methode.

I can do a PR if we agree on the method to be used everywhere: shell or regex?

### How can we reproduce this issue?

Step 1: Check for a string in a URL in the campaign decision: Clicks link with the filter `*test*`
Step 2: Check for a string in a URL in the campaign decision: Visited URL `*test*`
Step 3: Visited a URL will evaluate to true, Clicked a link will evaluate to false

### Relevant log output

_No response_

### Code of Conduct

- [X] I confirm that I have read and agree to follow this project's Code of Conduct

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://app.bountysource.com/issues/121828556-ux-bug-mautic-uses-different-ways-to-filter-urls-across-the-application-campaign-decisions-and-points?utm_campaign=plugin&utm_content=tracker%2F5355074&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://app.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F5355074&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by shinde-rahul**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | ✔️
| New feature/enhancement? (use the a.x branch)      | ❌
| Deprecations?                          | ❌
| BC breaks? (use the c.x branch)        | ❌
| Automated tests included?              | ✔️
| Related user documentation PR URL      | N/A
| Related developer documentation PR URL | N/A
| Issue(s) addressed                     | Fixes #12087

## Description

This PR standardizes URL filtering behavior across page-related campaign and point flows.

Before this change, Mautic used different matching rules depending on where the URL filter was used:
- email click decisions matched plain text as a normalized substring
- page campaign decisions used shell wildcards via `fnmatch`
- point URL actions used shell wildcards via `fnmatch`
- lead page-hit conditions used shell wildcards via `fnmatch`

That made the same filter behave differently across the UI. For example, `test` matched email click URLs but did not match page visit or points URL filters unless the user entered `*test*`.

This PR updates the remaining page-related flows to use `UrlMatcher` so plain text filters now behave consistently as substring matches across the application. To avoid breaking existing saved configurations, legacy wildcard patterns such as `*test*` are still supported.

---
### 📋 Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create or open a campaign and add the `Visits a page` decision.
3. Make sure the contact is already in the campaign before testing the page hit.
4. In the `URL` field, enter `test` and save the event.
5. Trigger a tracked page hit with a URL that contains `test`, for example, `https://mautic.ddev.site/my-test-page`.
6. Confirm the campaign decision evaluates to true.
7. Edit the same decision and replace `test` with `*test*`.
8. Trigger the same tracked page hit again and confirm the decision still evaluates to true.
9. Repeat the same validation using the `Referrer` field with a referrer URL that contains `test`.
10. Create a Points action using `Visits specific URL`.
11. Enter `test` in the `Page URL` field and trigger a tracked page hit with a URL containing `test`.
12. Confirm the points action is applied.
13. Change the same filter to `*test*` and confirm it still applies.
14. Create or edit a campaign condition using the lead page-hit URL filter.
15. Enter `test` and confirm a contact with a page-hit URL containing `test` matches the condition.
16. Change the filter to `*test*` and confirm the same contact still matches.
17. Verify a non-matching filter such as `does-not-match` does not match in the page campaign decision, point URL action, and lead page-hit condition flows.

### CODE REVIEW

**Comment by promptless-for-oss** [comment]
Promptless prepared documentation updates related to this change.

**User Documentation** ([mautic/user-documentation](https://github.com/mautic/user-documentation)):
- Updated URL matching documentation in campaign troubleshooting, points troubleshooting, and campaign builder
- [Review suggestion](https://app.gopromptless.ai/suggestions/38a3130e-89d2-47fa-ba9d-22dd70b3b586)

**Developer Documentation** ([mautic/developer-documentation-new](https://github.com/mautic/developer-documentation-new)):
- Added `UrlMatcher` helper class documentation with API reference and usage examples
- [Review suggestion](https://app.gopromptless.ai/suggestions/18665108-2e2f-4a23-a856-bf9998461374)

Summary: Documents the standardized URL filter matching behavior that now works consistently as substring matching across all campaign decisions and point actions.

**Comment by andersonjeccel** [comment]
@shinde-rahul do you plan to address sonarqube?

**Comment by andersonjeccel** [comment]
@shinde-rahul CI found a few failures. Do you have a minute to review them?

**Review by hluchas — APPROVED** [review]
Looks good and fix works.
Could you please fix that PHPStan issue? https://github.com/mautic/mautic/actions/runs/27626885014/job/81690839251?pr=15979
TY!

File: app/bundles/EmailBundle/Helper/UrlMatcher.php [hunk 1]
```diff
@@ -5,15 +5,18 @@
...
-    public static function hasMatch(array $urlsToCheckAgainst, $urlToFind): bool
+    public static function hasMatch(array $urlsToCheckAgainst, mixed $urlToFind): bool
     {
-        $urlToFind = self::sanitizeUrl($urlToFind);
+        $urlToFind = self::sanitizeUrl((string) $urlToFind);
 
         foreach ($urlsToCheckAgainst as $url) {
-            $url = self::sanitizeUrl($url);
+            $url = self::sanitizeUrl((string) $url);
+
```
Comment (andersonjeccel) [inline app/bundles/EmailBundle/Helper/UrlMatcher.php:22]: one edge case: an empty filter value now matches every url because it reaches `preg_match(//i, $urlToFind)`. a trailing comma in the campaign url field can produce that empty value, so we should skip it before matching.

**Comment by shinde-rahul** [comment]
@andersonjeccel Please re-review. @hluchas, the CI errors are fixed now.

File: app/bundles/EmailBundle/Helper/UrlMatcher.php (hunk 1 above)
Comment (shinde-rahul) [reply app/bundles/EmailBundle/Helper/UrlMatcher.php:22]: Please review.

File: app/bundles/EmailBundle/Tests/Helper/UrlMatcherTest.php [hunk 2]
```diff
@@ -72,6 +72,36 @@
         $this->assertFalse(UrlMatcher::hasMatch($urls, 'https://yahoo.com'));
     }
 
+    public function testEmptyUrlFilterIsIgnored(): void
+    {
+        $urls = [
+            'http://google.com',
```
[the remark this answers is not in the document]
Comment (shinde-rahul) [reply app/bundles/EmailBundle/Tests/Helper/UrlMatcherTest.php:80]: This is an example for testing, and safe to ignore.

**Comment by shinde-rahul** [comment]
SonarQube error is safe to ignore. This is yielding on an edge-case data set.

**Comment by andersonjeccel** [comment]
@shinde-rahul Some tests failed in CI. Could you check them?

**Review by andersonjeccel — APPROVED** [review]
no concerns here

**Review by andersonjeccel — APPROVED** [review]
Pass, tested:

- triggered browser page hit with test
- point action matched test
- changed URL to *test*
- point action matched wildcard
- campaign matched both

**Comment by shinde-rahul** [comment]
@andersonjeccel and @hluchas, if you can spare some time to look at this again, I have resolved the conflicts.

File: app/bundles/EmailBundle/Helper/UrlMatcher.php [hunk 3]
```diff
@@ -5,16 +5,25 @@
 class UrlMatcher
 {
     /**
-     * @param string[] $urlsToCheckAgainst
-     * @param string   $urlToFind
+     * @param array<array-key, mixed> $urlsToCheckAgainst
      */
-    public static function hasMatch(array $urlsToCheckAgainst, $urlToFind): bool
+    public static function hasMatch(array $urlsToCheckAgainst, mixed $urlToFind): bool
```
Comment (escopecz) [inline app/bundles/EmailBundle/Helper/UrlMatcher.php]: why the change to mixed?

**Review by hluchas — APPROVED** [review]

File: app/bundles/EmailBundle/Helper/UrlMatcher.php [hunk 4]
```diff
@@ -5,16 +5,25 @@
 class UrlMatcher
 {
     /**
-     * @param string[] $urlsToCheckAgainst
-     * @param string   $urlToFind
+     * @param array<array-key, mixed> $urlsToCheckAgainst
```
Comment (kniziol) [inline app/bundles/EmailBundle/Helper/UrlMatcher.php]: suggestion: https://phpstan.org/writing-php-code/phpdoc-types#array-shapes

File: app/bundles/EmailBundle/Helper/UrlMatcher.php (hunk 3 above)
Comment (kniziol) [inline app/bundles/EmailBundle/Helper/UrlMatcher.php]:
```suggestion
    public static function hasMatch(array $urlsToCheckAgainst, string $urlToFind): bool
```
Will it work?

File: app/bundles/EmailBundle/Helper/UrlMatcher.php [hunk 5]
```diff
@@ -5,16 +5,25 @@
...
         foreach ($urlsToCheckAgainst as $url) {
+            $url = (string) $url;
+
+            if ('' === $url) {
+                continue;
+            }
+
             $url = self::sanitizeUrl($url);
 
+            if (self::isLegacyWildcardPattern($url) && fnmatch($url, $urlToFind, FNM_CASEFOLD)) {
```
Comment (kniziol) [inline app/bundles/EmailBundle/Helper/UrlMatcher.php]:
```suggestion
            if (self::isLegacyWildcardPattern($url) && \fnmatch($url, $urlToFind, FNM_CASEFOLD)) {
```

File: app/bundles/EmailBundle/Helper/UrlMatcher.php [hunk 6]
```diff
@@ -48,4 +57,9 @@ private static function sanitizeUrl(string $url): string
 
         return $url;
     }
+
+    private static function isLegacyWildcardPattern(string $url): bool
+    {
+        return false !== strpbrk($url, '*?[');
```
Comment (kniziol) [inline app/bundles/EmailBundle/Helper/UrlMatcher.php]:
```suggestion
        return false !== \strpbrk($url, '*?[');
```

File: app/bundles/PageBundle/EventListener/CampaignSubscriber.php [hunk 7]
```diff
@@ -180,32 +181,24 @@ public function onCampaignTriggerDecision(CampaignExecutionEvent $event): bool|C
...
         // Check Landing Pages URL or Tracing Pixel URL
         if (isset($config['url']) && $config['url']) {
             $pageUrl     = html_entity_decode($eventDetails->getUrl());
             $limitToUrls = explode(',', $config['url']);
-
-            foreach ($limitToUrls as $url) {
-                $url              = html_entity_decode(trim($url));
-                $urlMatches[$url] = fnmatch($url, $pageUrl);
-            }
+            $limitToUrls = array_map(static fn (string $url): string => html_entity_decode(trim($url)), $limitToUrls);
```
Comment (kniziol) [inline app/bundles/PageBundle/EventListener/CampaignSubscriber.php]:
```suggestion
            $limitToUrls = \array_map(static fn (string $url): string => \html_entity_decode(\trim($url)), $limitToUrls);
```

File: app/bundles/PageBundle/EventListener/CampaignSubscriber.php [hunk 8]
```diff
@@ -180,32 +181,24 @@ public function onCampaignTriggerDecision(CampaignExecutionEvent $event): bool|C
...
         // Check Landing Pages URL or Tracing Pixel URL
         if (isset($config['referer']) && $config['referer']) {
             $refererUrl      = html_entity_decode($eventDetails->getReferer());
             $limitToReferers = explode(',', $config['referer']);
-
-            foreach ($limitToReferers as $referer) {
-                $referer                  = html_entity_decode(trim($referer));
-                $refererMatches[$referer] = fnmatch($referer, $refererUrl);
-            }
+            $limitToReferers = array_map(static fn (string $referer): string => html_entity_decode(trim($referer)), $limitToReferers);
```
Comment (kniziol) [inline app/bundles/PageBundle/EventListener/CampaignSubscriber.php]:
```suggestion
            $limitToReferers = \array_map(static fn (string $referer): string => \html_entity_decode(\trim($referer)), $limitToReferers);
```

File: app/bundles/PageBundle/Helper/PointActionHelper.php [hunk 9]
```diff
@@ -101,4 +102,13 @@ public function validateUrlHit($eventDetails, array $action): bool
         // return true only if all configured options are true
         return !in_array(false, $changePoints);
     }
+
+    private function getSqlLikePattern(string $url): string
+    {
+        if (false !== strpbrk($url, '*?')) {
```
Comment (kniziol) [inline app/bundles/PageBundle/Helper/PointActionHelper.php]:
```suggestion
        if (false !== \strpbrk($url, '*?')) {
```

File: app/bundles/PageBundle/Helper/PointActionHelper.php [hunk 10]
```diff
@@ -101,4 +102,13 @@ public function validateUrlHit($eventDetails, array $action): bool
...
+            return str_replace(['*', '?'], ['%', '_'], $url);
```
Comment (kniziol) [inline app/bundles/PageBundle/Helper/PointActionHelper.php]:
```suggestion
            return \str_replace(['*', '?'], ['%', '_'], $url);
```

File: app/bundles/PageBundle/Tests/Helper/PointActionHelperFunctionalTest.php [hunk 11]
```diff
@@ -156,6 +156,64 @@ public function testValidateUrlHitWithoutPreviousHit(): void
...
+            'product/123',
+            'Plain text URL filters should match persisted hits via repository lookups.',
+        ];
+        yield 'legacy wildcard' => [
+            '*product/123*',
+            'Legacy wildcard URL filters should still match persisted hits via repository lookups.',
+        ];
+    }
+
+    #[\PHPUnit\Framework\Attributes\DataProvider('repositoryUrlFilterProvider')]
```
Comment (kniziol) [inline app/bundles/PageBundle/Tests/Helper/PointActionHelperFunctionalTest.php]:
```suggestion
    #[DataProvider('repositoryUrlFilterProvider')]
```
plus at the top:
```php
use PHPUnit\Framework\Attributes\DataProvider;
```

**Review by kniziol — APPROVED** [review]
It looks great. I’ve left a few suggestions.

File: app/bundles/PageBundle/Helper/PointActionHelper.php [hunk 12]
```diff
@@ -101,4 +102,13 @@ public function validateUrlHit($eventDetails, array $action): bool
...
+        }
+
+        return '%'.$url.'%';
+    }
```
Comment (andersonjeccel) [inline app/bundles/PageBundle/Helper/PointActionHelper.php:114]: A plain-text filter containing `_` or `%` matches literally in `UrlMatcher`, but this passes them to SQL `LIKE` as wildcards. Can we escape them before adding the surrounding `%`?

File: app/bundles/EmailBundle/Helper/UrlMatcher.php (hunk 4 above)
Comment (shinde-rahul) [reply app/bundles/EmailBundle/Helper/UrlMatcher.php]: Done, thanks.

File: app/bundles/EmailBundle/Helper/UrlMatcher.php (hunk 3 above)
Comment (shinde-rahul) [reply app/bundles/EmailBundle/Helper/UrlMatcher.php]: No strong reason. `mixed` only preserved the previous loose input behavior, but callers pass URL strings, so I’ve changed it to `string`.

File: app/bundles/PageBundle/Helper/PointActionHelper.php (hunk 12 above)
Comment (shinde-rahul) [reply app/bundles/PageBundle/Helper/PointActionHelper.php:114]: Updated code and added a designated test case.

**Comment by adiati98** [comment]
Hi @shinde-rahul,

Promptless created PRs at both user and developer docs based on the changes in this PR.
Can you please review the contents and see if they're align with your changes? TIA! ✨ 

https://github.com/mautic/user-documentation/pull/919
